### PR TITLE
Fix eslint issues in app/javascript/listings

### PR DESCRIPTION
**Pull request by hoangvvo**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description

Fix some eslint (some PropTypes, destructuring, etc.) and eslint a11y issues as well as modifying the test to have the new a11y

## Related Tickets & Documents

#2470

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

Only some a11y additions to buttons, and those with onClick

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

reddd error everywhere

### CODE REVIEW

**Comment by hoangvvo** [comment]
That codeclimate check is so 💩 @@

File: app/javascript/listings/elements/categories.jsx [hunk 1]
```diff
@@ -3,62 +3,66 @@ import { h, Component } from 'preact';
...
   }
 }
 
 Categories.propTypes = {
-  categoriesForSelect: PropTypes.array.isRequired,
-  categoriesForDetails: PropTypes.array.isRequired,
+  categoriesForSelect: PropTypes.arrayOf(PropTypes.string).isRequired,
+  categoriesForDetails: PropTypes.arrayOf(
+    PropTypes.shape({
+      name: PropTypes.string,
```
Comment (hoangvvo) [inline app/javascript/listings/elements/categories.jsx]: Please correct me if I'm wrong in this propTypes

File: app/javascript/listings/listingForm.jsx [hunk 2]
```diff
@@ -119,3 +125,10 @@ export default class ListingForm extends Component {
     );
   }
 }
+
+ListingForm.propTypes = {
```
Comment (hoangvvo) [inline app/javascript/listings/listingForm.jsx]: Please correct me if I'm wrong in this propTypes also..

File: app/javascript/listings/singleListing.jsx [hunk 3]
```diff
@@ -82,7 +92,18 @@ export const SingleListing = ({
 };
 
 SingleListing.propTypes = {
-  listing: PropTypes.object.isRequired,
+  listing: PropTypes.shape({
```
Comment (hoangvvo) [inline app/javascript/listings/singleListing.jsx]: Please correct me if I'm wrong in this propTypes also..

**Comment by hoangvvo** [comment]
image
Done ✔️ 
Onl blocked with some unreasonable code climate checks.

File: app/javascript/listings/listings.jsx [hunk 4]
```diff
@@ -310,6 +342,9 @@ export class Listings extends Component {
             className="tag-close"
             onClick={e => this.removeTag(e, tag)}
             data-no-instant
+            role="button"
+            onKeyPress={e => this.removeTag(e, tag)}
```
Comment (rhymes) [inline app/javascript/listings/listings.jsx]: cool!

File: app/javascript/listings/listings.jsx [hunk 5]
```diff
@@ -360,21 +395,30 @@ export class Listings extends Component {
           id="classified-listings-modal-background"
         />
       );
-      if (
-        openedListing.contact_via_connect &&
-        openedListing.user_id !== currentUserId
```
Comment (rhymes) [inline app/javascript/listings/listings.jsx]: why was this removed and moved down?

File: app/javascript/listings/singleListing.jsx [hunk 6]
```diff
@@ -82,7 +92,18 @@ export const SingleListing = ({
...
+    id: PropTypes.number,
+    category: PropTypes.string,
+    contact_via_connect: PropTypes.bool,
+    location: PropTypes.string,
+    processed_html: PropTypes.string,
+    slug: PropTypes.string,
+    title: PropTypes.string,
+    user_id: PropTypes.number,
+    tag_lis: PropTypes.array,
```
Comment (rhymes) [inline app/javascript/listings/singleListing.jsx]: there's a typo here, it should be `tag_list`, also possibly `arrayOf(string)`

File: app/javascript/listings/singleListing.jsx [hunk 7]
```diff
@@ -82,7 +92,18 @@ export const SingleListing = ({
...
+    author: PropTypes.object,
```
Comment (rhymes) [inline app/javascript/listings/singleListing.jsx]: this should be a shape object with `name`, `username` and `profile_image_90` according to the logs

**Comment by rhymes** [comment]
@hoangvvo there's also a conflict in `categories.jsx` because a previous PR - #4257 - has been merged to clear the names of the array params

File: app/javascript/listings/elements/categories.jsx [hunk 8]
```diff
@@ -3,62 +3,66 @@ import { h, Component } from 'preact';
...
   options = () => {
-    const { categoriesForSelect, category } = this.props
+    const { categoriesForSelect, category } = this.props;
     return categoriesForSelect.map(array => {
       // array example: ["Education/Courses (1 Credit)", "education"]
-      if(category === array[1]) {
-        return(
-          <option value={array[1]} selected>{array[0]}</option>
-        )
+      if (category === array[1]) {
```
[the remark this answers is not in the document]
Comment (nickytonline) [reply app/javascript/listings/elements/categories.jsx]: The destructuring is already handled in this merged PR, https://github.com/thepracticaldev/dev.to/pull/4257.

File: app/javascript/listings/elements/categories.jsx [hunk 9]
```diff
@@ -3,62 +3,66 @@ import { h, Component } from 'preact';
...
+    );
+  };
 
   render() {
-    const { onChange } = this.props
-    return(
+    const { onChange } = this.props;
+    return (
       <div className="field">
         <label className="listingform__label" htmlFor="category">
```
Comment (ruffle1986) [inline app/javascript/listings/elements/categories.jsx]: Sorry if I'm missing something here but it says it's a label for an element with id "category" (htmlFor="category") but there's no element with id category.

File: app/javascript/.eslintrc.js [hunk 10]
```diff
@@ -23,6 +23,11 @@ module.exports = {
     ],
     'import/prefer-default-export': 'off',
     'no-unused-vars': ['error', { argsIgnorePattern: '^_' }],
+    'jsx-a11y/label-has-associated-control': [2, {
```
Comment (ruffle1986) [inline app/javascript/.eslintrc.js]: I think being consistent and explicit is a good thing. I would write 'error' instead of the number. Like how it is on the previous line.

File: app/javascript/listings/elements/expireDate.jsx [hunk 11]
```diff
@@ -5,13 +5,16 @@ const ExpireDate = ({ onChange, defaultValue }) => {
   let tomorrow = new Date();
   let monthFromToday = new Date();
   tomorrow.setDate(new Date().getDate() + 1);
-  tomorrow = tomorrow.toISOString().split('T')[0];
+  [tomorrow] = tomorrow.toISOString().split('T');
```
[the remark this answers is not in the document]
Comment (hoangvvo) [reply app/javascript/listings/elements/expireDate.jsx]: @rhymes Yes, I don't really want to do that, but Eslint was complaining...

File: app/javascript/listings/listings.jsx (hunk 5 above)
Comment (hoangvvo) [reply app/javascript/listings/listings.jsx]: I tried to avoid code duplicates in that part.
The original one repeat the whole <form> when the only difference are the two texts.

The condition contains `openedListing.contact_via_connect` in either case so I move it up, and also for me to use the `? ... : ...` syntax.

File: app/javascript/listings/singleListing.jsx (hunk 6 above)
Comment (hoangvvo) [reply app/javascript/listings/singleListing.jsx]: @rhymes thanks for the catch

File: app/javascript/listings/elements/categories.jsx (hunk 9 above)
Comment (hoangvvo) [reply app/javascript/listings/elements/categories.jsx]: I will assume it is the <select> after that. Interesting that Eslint did not pick that up

File: app/javascript/listings/listings.jsx (hunk 5 above)
Comment (hoangvvo) [reply app/javascript/listings/listings.jsx]: I was trying to remove duplicate. The whle <form> is repeated even though the only differences are the two texts

Both case of the condition contains `openedListing.contact_via_connect` so I put that to the top, and that also allows me to use ` ? ... : ...` syntax.

File: app/javascript/listings/listings.jsx (hunk 4 above)
Comment (ruffle1986) [reply app/javascript/listings/listings.jsx]: removeTag should be called when the user presses enter:

`e.key === 'Enter'`

File: app/javascript/listings/listingDashboard.jsx [hunk 12]
```diff
@@ -110,7 +110,7 @@ export class ListingDashboard extends Component {
         className={`rounded-btn ${filter === f ? 'active' : ''}`}
         role="button"
         onKeyPress={event => {
-          this.setState({ filter: event.target.textContent });
+          event.key === 'Enter' && this.setState({ filter: event.target.textContent });
```
[the remark this answers is not in the document]
Comment (hoangvvo) [reply app/javascript/listings/listingDashboard.jsx]: Sure, that will do. I will make the changes shortly.

**Comment by hoangvvo** [comment]
Screenshot from 2019-10-09 16-34-56

Funky. Husky hook keep making this change when I commit, causing the test to fail. (At `Credits Available`)

File: app/javascript/listings/listingDashboard.jsx [hunk 13]
```diff
@@ -109,9 +109,10 @@ export class ListingDashboard extends Component {
         }}
         className={`rounded-btn ${filter === f ? 'active' : ''}`}
         role="button"
-        onKeyPress={event => {
-          event.key === 'Enter' && this.setState({ filter: event.target.textContent });
-        }}
+        onKeyPress={event =>
+          (event.key === 'Enter' || event.key === ' ') &&
+          this.setState({ filter: event.target.textContent })
+        }
```
Comment (rhymes) [inline app/javascript/listings/listingDashboard.jsx]: could you use a function for these three event handlers? the only difference is the hash of the state
